### DISCUSSION TITLE
Custom action in InstallUISequence

### DISCUSSION THREAD

**Discussion by bodnarpolina** [oleg-shilo/wixsharp#1458] [Q&A] [answered]
Hello, @oleg-shilo! We are using _WixSharp_ in our project. I know how to create a custom action and add it to _InstallExecuteSequence_. Usually we use _ManagedAction_ or _ElevatedManagedAction_. Everything works fine! But is it a way to add a custom action to the _InstallUISequence_? I've found [this](https://github.com/oleg-shilo/wixsharp/wiki/Deployment-scenarios#embedded-ui-limitations) section in the documentation, but I'm not quite sure it's what I need. Can you please help me?

**Comment by oleg-shilo** [comment]
TBH my by far preferred approach is to completely abandon raw WiX custom actions and use managed actions that are wrapped int the `ManagedProject` events. It so-o-o much simpler and more straight forward. There are only three simple events that are naturally mapped to the setup steps (mot some tables in MSI 😄 ): `Load`, `BeforeInstall` and `AfterInstall`. And the `AfterInstall` is already elevated.

You can read more here: https://github.com/oleg-shilo/wixsharp/wiki/Managed-Setup-Model#event-driven-custom-actions 

You see, MSI/WiX are very limited in giving the user the ability to define even a moderate complexity algorithm so they force the user to create custom actions for everything. You want to check registry - create a custom action, check file system - another custom action. replace a value in the json file - another custom action. All of them are running as separate process and you cannot easily pass the data between them. And debugging it is a complete nightmare.

While all this can be done in three lines of code in C#/Python/Java/JS (any proper language), in MSI world you need to do it as it was common in 1990.

That's why I strongly encourage you to move your code to events of the `ManagedProject`. When you run managed UI you will have another two events associated with the UI initialization. 

You can find plenty of examples in the [Managed Setup](https://github.com/oleg-shilo/wixsharp/tree/master/Source/src/WixSharp.Samples/Wix%23%20Samples/Managed%20Setup) folder.

The setups are deliberately saturated with various scenarios so they can demonstrate all features. But in reality, you will be using only a fraction of what you see there.

Let me know if it helps and if you need any guidance with that.

**Reply by bodnarpolina** [reply]
@oleg-shilo , thanx a lot for your detailed answer! I have read all the links and articles and tried to apply it to my current task. But with no success, unfortunately. I'm gonna explain what I have to do. In fact I need to run a standard action _FindRelatedProducts_ one more time. Like in [this](https://stackoverflow.com/a/35064434) answer. At first I tried to run a _ManagedAction_. Here is a piece of my code:
```
public static ActionResult FindRelatedProductsActionForUserScope(Session session)
    {
      session.Log("Begin FindRelatedProductsActionForUserScope");
      
      try
      {
        //switching to the user scope
        session[AllUsersProperty] = "";
        session.Log($"ALLUSERS property is set to {session[AllUsersProperty]}");
        session.DoAction(FindRelatedProductsAction);
        //switching back to the machine scope
        session[AllUsersProperty] = "1";
      }
      catch (Exception ex)
      {
        session.Log($"Performing \"FindRelatedProducts\" action for the user scope has failed due to \"{ex.GetType().Name}\": \"{ex.Message}\"");
        return ActionResult.Failure;
      }
      
      return ActionResult.Success;
    }
```

And this action is successfully run in _InstallExecuteSequence_. But with no luck :(. Because it was already run in _InstallUiSequence_.! So I usually get a message in my logs: _"Skipping FindRelatedProducts action: already done on client side"_. That's why I had an idea to run this action in _InstallUiSequence_ for the second time. I tried to use the links above, but still stuck a bit. According to the documentation I can use _project.UIInitialized_ or _project.UILoaded_ only if I have _project.ManagedUI_ set. But we don't have it in  our ManagedProject as we use `UI = WUI.WixUI_ProgressOnly` and I don't want to change this behavior and add any new features. So, maybe in our case it's impossible to run the above piece of code on _project.UIInitialized_ or _project.UILoaded_ at all?

P.S. Maybe there is another completely different way to run _FindRelatedProducts_ for the second time?

**Comment by oleg-shilo** [comment]
Sorry, but this is exactly what I was trying to avoid with my suggestion. 😄 

> I'm gonna explain what I have to do. In fact I need to run a standard action FindRelatedProducts one more time.

Forget about WiX/MSI for a minute and try to think about it from the "what I want to achieve" angle. Most likely you want to understand if something is installed or not. But not to execute a specific algorithm in the intricate runtime architecture of MSI.

Thus all your code above can be collapsed into this:

```C#
project.BeforeInstall += args =>
{
    string[] productCodes = AppSearch.GetRelatedProducts(project.UpgradeCode.ToString());
    // do whatever you want to do with this info
```

You have the power of C#, it's much easier to do it by yourself directly where you need it instead of trying to invoke multiple C++ routines scattered around WiX dll(s) and also fiddling with the session object.

You can also look at the implementation of "AppSerach.GetRelatedProducts". It simply calls MSI's `MsiEnumRelatedProducts`.

```C#
 [DllImport("msi")]
 static extern int MsiEnumRelatedProducts(string productCode, int reserved, int iProductIndex, StringBuilder lpProductBuf);
```
And most likely there will be some extra context that you can pass to the call if required.

Hope this is what you were looking for.

**Reply by bodnarpolina** [reply]
@oleg-shilo , thank you very much for your ideas! I tried them all, but again faced a problem. What I need to do in general is simply to remove all of the apps that was installed before the current installation. In general it's all supported by msi means: upgrade policies, _FindRelatedProductsAction_ and _RemoveExistingProductsAction_. But not for my case :( As I wrote before we switched a _user scope_ to _machine scope_ simply adding `InstallScope = InstallScope.perMachine`. And as a result _FindRelatedProductsAction_ skips the previous installed apps as they were installed for the user scope! Like this: _"FindRelatedProducts: current install is per-machine.  Related install for product '{0C6604FB-58EC-48B9-8259-5871EFDADEB9}' is per-user.  Skipping..."_. That's why I provided [this](https://stackoverflow.com/questions/678002/how-do-i-fix-the-upgrade-logic-of-a-wix-setup-after-changing-installscope-to-pe/35064434#35064434) link above. I guess I should have started with this info asking my first question :) That's why we had an idea of running _FindRelatedProductsAction_ as a custom action separately for a user scope. Getting back to the idea you described in the last answer: I've found all installed apps with `AppSearch.GetRelatedProducts(project.UpgradeCode.ToString())` successfully! But when trying to uninstall them with a command like `Installer.ConfigureProduct("{MY-PRODUCT-GUID}", 0, InstallState.Absent, "");` I get an exception saying that I cannot run one more installation until I finish the current one: _"Another Installation In Progress Please Complete Installation Before Continuing"_. I tried to move the piece of code from _BeforeInstall_ to _AfterInstall_ thinking it would solve a problem but still got the same error :(

P.S. I've violated your advice about running _FindRelatedProductsAction_ as a custom action in _UIInitialized_ section just for testing purposes :) As a result I couldn't run it with the exception like: "Handle is in an invalid state." My code was
```
project.ManagedUI = new ManagedUI();
project.ManagedUI.InstallDialogs.Add(Dialogs.Progress)
                                       .Add(Dialogs.Exit);
project.UIInitialized += (SetupEventArgs e) =>
      {
          e.Session[AllUsersProperty] = "";
          e.Session.DoAction(FindRelatedProductsAction);
     }
```
Is this what you have warned me about saying "Sorry, but this is exactly what I was trying to avoid with my suggestion." ? :)

**Comment by oleg-shilo** [comment, marked as answer]
Correct. You need to do these actions before your session starts. In the `pre-install` phase:
![288575163-efbb4bbc-eacb-4de9-8f3f-d9b81677fbf1](https://github.com/oleg-shilo/wixsharp/assets/16729806/5e8ab204-f195-469f-8512-9f2201ca22f5)
